test/test-agent.cpp

```diff
@@ -0,0 +1,69 @@
+#include <gtest/gtest.h>
+
+#include "agent.h"
+
+TEST(AgentTest, Basic)
+{
+    Rep::Agent agent = Rep::Agent().allow("/").disallow("/foo");
+    EXPECT_EQ(agent.directives().size(), 2);
+}
+
+TEST(AgentTest, ChecksAllowed)
+{
+    Rep::Agent agent = Rep::Agent().allow("/path");
+    EXPECT_TRUE(agent.allowed("/path"));
+    EXPECT_TRUE(agent.allowed("/elsewhere"));
+}
+
+TEST(AgentTest, HonorsLongestFirstPriority)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/path").allow("/path/exception");
+    EXPECT_TRUE(agent.allowed("/path/exception"));
+    EXPECT_FALSE(agent.allowed("/path"));
+}
+
+TEST(AgentTest, RobotsTextAllowed)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/robots.txt");
+    EXPECT_TRUE(agent.allowed("/robots.txt"));
+}
+
+TEST(AgentTest, DisallowNone)
+{
+    Rep::Agent agent = Rep::Agent().disallow("");
+    EXPECT_TRUE(agent.allowed("/anything"));
+}
+
+TEST(AgentTest, EscapedRule)
+{
+    Rep::Agent agent = Rep::Agent().disallow("/a%3cd.html");
```

Comment:
Question: Do you normalize to the unescaped rule and if so how does this interact with the longest rule match criteria. It seems like an escaped rule might need to overtake an unescaped rule due to length, but I'm not sure how that actually plays out and if this edge case is actually referenced anywhere.